client/src/main/java/hudson/plugins/swarm/Client.java

```diff
@@ -47,12 +49,28 @@ public static void main(String... args) throws InterruptedException {
             System.exit(0);
         }
 
+        if (options.config != null) {
+            logger.log(Level.INFO, "Load configuration from {0}", options.config.getPath());
+
+            try (InputStream is = new FileInputStream(options.config)) {
+                options = loadFromConfig(is);
+            } catch (IOException ex) {
+                throw new UncheckedIOException("Failed to load configuration", ex);
+            }
+        }
+
         validateOptions(options);
 
         // Pass the command line arguments along so that the LabelFileWatcher thread can have them.
         run(new SwarmClient(options), options, args);
     }
 
+    static Options loadFromConfig(InputStream is)
+    {
+        final Yaml yaml = new Yaml();
```

Comment:
I'm not sure about what the current recommendation is, but at least a few years ago there were some concerns over parsing untrusted input with a regular YAML parser like this. For example, a user could refer to a Java class in the YAML file and trigger unexpected execution of its constructor.

The [SnakeYAML documentation](https://bitbucket.org/asomov/snakeyaml/wiki/Documentation) writes:

> _Note:_ if you want **to limit** objects to standard Java objects like `List` or `Long` you need [to use `SafeConstructor`](https://bitbucket.org/asomov/snakeyaml/src/master/src/test/java/examples/SafeConstructorExampleTest.java).

```java
Yaml yaml = new Yaml(new SafeConstructor());
```

This would require further changes to the parsing logic to avoid the use of custom types, but I think it would make the implementation more bulletproof to unexpected user input and possibly keep the [Jenkins security team](https://www.jenkins.io/security/team/) from knocking on my door in the future. :-) It might be best to tackle this alongside input validation.

Reply:
Thanks, I'll have a look at it.

Regarding input validation: Where is this done best? Manual code directly after deserialization (the code part highlighted by the comment)?

Reply:
Input validation added. The rules are equal to CLI with two exceptions:  `-config` and `-help` are not allowed as they do not make sense here.

I have added a `Constructor` limited to the specific class (as recommended [here](https://securitylab.github.com/research/swagger-yaml-parser-vulnerability/)) to mitigate the unsafe Yaml. In doubt, can we get someone of the security team having a look at it?